### PR TITLE
refactor: updating demo-client to the latest SDK version

### PR DESCRIPTION
**Pull request by Dudleyneedham**
## fixes https://github.com/KILTprotocol/ticket/issues/1045
Update the sdk in the demo-client

## How to test:
Use the dev tools and then try to run through the application. 

## Checklist:

- [ ] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: src/components/DelegationNode/DelegationNode.tsx [hunk 1]
```diff
@@ -283,16 +290,41 @@ class DelegationNode extends React.Component<Props, State> {
...
+    if (firstAttestation === null) {
+      throw SDKErrors.ERROR_NOT_FOUND('Attestation not on chain')
+    }
+
+    const steps = await DelegationsService.checkTraversalStepsToParent(
+      selectedIdentity.identity,
+      firstAttestation
+    )
+
+    await Promise.chain(
```
Comment (rflechtner) [inline src/components/DelegationNode/DelegationNode.tsx:304]: I didn't know this existed! how does this compare to `Promise.all()` ?

File: src/services/DelegationsService.ts [hunk 2]
```diff
@@ -172,6 +175,42 @@ class DelegationsService {
 
     return delegation.store(selectedIdentity)
   }
+
+  public static async checkTraversalStepsToParent(
+    attester: Identity,
+    attestation: Attestation
```
Comment (rflechtner) [inline src/services/DelegationsService.ts:181]: I'm surprised we don't have that in the SDK

**Review by rflechtner — APPROVED** [review]
LGTM, although some of the traversal steps related code may be moved into the sdk

File: src/components/DelegationNode/DelegationNode.tsx (hunk 1 above)
Comment (Dudleyneedham) [reply src/components/DelegationNode/DelegationNode.tsx:304]: It has an option to continue through the promises even after a promise fails for whatever reason. 

If three promises are chained and the middle one throws an error it won't stop and throw and error straight away. It will continue to the next one and check if it is a success or not, then it will parse the successes and failures to a follow-up `then` to show errors or successful outcomes. 

During the function, it will complete the action if successful.


I hope I explained it well enough XD

File: src/services/DelegationsService.ts (hunk 2 above)
Comment (Dudleyneedham) [reply src/services/DelegationsService.ts:181]: We do now!

File: src/components/DelegationDetailView/DelegationDetailView.tsx [hunk 3]
```diff
@@ -1,9 +1,8 @@
 import {
-  DelegationNode as SDKDelegationNode,
   DelegationBaseNode,
-  IDelegationBaseNode,
-  IDelegationRootNode,
+  DelegationNode as SDKDelegationNode,
 } from '@kiltprotocol/sdk-js'
+import { IDelegationBaseNode, IDelegationRootNode } from '@kiltprotocol/types'
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/components/DelegationDetailView/DelegationDetailView.tsx:5]: Added

File: src/components/DevTools/DevTools.wallet.tsx [hunk 4]
```diff
@@ -35,7 +35,7 @@ class BsIdentity {
 
   public static async create(alias: string): Promise<void | IMyIdentity> {
```
Comment (tjwelde) [inline src/components/DevTools/DevTools.wallet.tsx:36]: Is this still async now?

File: src/components/MessageDetailView/MessageDetailView.tsx [hunk 5]
```diff
@@ -128,9 +129,12 @@ class MessageDetailView extends React.Component<Props, State> {
         )
       }
       case MessageBodyType.SUBMIT_TERMS: {
+        // Need to fix with the update in Message Compresss and Decompress
```
Comment (tjwelde) [inline src/components/MessageDetailView/MessageDetailView.tsx:132]: What does this mean?

File: src/services/AttestationWorkflow.ts [hunk 6]
```diff
@@ -137,22 +137,22 @@ class AttestationWorkflow {
       throw new Error('No selected Identity')
     }
 
-    const requestForAttestation = await RequestForAttestation.fromClaimAndIdentity(
+    const requestForAttestation = RequestForAttestation.fromClaimAndIdentity(
```
Comment (tjwelde) [inline src/services/AttestationWorkflow.ts:140]: Is the function `submitClaimsForCTypes` now still async?

File: src/containers/WalletAdd/WalletAdd.tsx [hunk 7]
```diff
@@ -72,7 +72,7 @@ class WalletAdd extends React.Component<Props, State> {
     let identity
     const phrase = useMyPhrase ? myPhrase : randomPhrase
     try {
-      identity = await Identity.buildFromMnemonic(phrase)
+      identity = Identity.buildFromMnemonic(phrase)
```
Comment (tjwelde) [inline src/containers/WalletAdd/WalletAdd.tsx:75]: `addIdentity` still async?

File: src/components/SelectClaims/SelectClaims.tsx [hunk 8]
```diff
@@ -18,7 +18,7 @@ type SelectOption = {
 type Props = RouteComponentProps<{}> & {
   closeMenuOnSelect?: boolean
   claims?: Claims.Entry[]
-  cTypeHash: ICType['cType']['hash'] | null
+  cTypeHash: ICType['ctypeHash'] | null
```
Comment (tjwelde) [inline src/components/SelectClaims/SelectClaims.tsx:21]: `ICType` here is coming from `'../../types/Ctype'`. are you sure about this change?

File: src/services/MessageRepository.tsx [hunk 9]
```diff
@@ -246,7 +247,11 @@ class MessageRepository {
       case MessageBodyType.REQUEST_TERMS:
         return [(message.body as IRequestTerms).content.cTypeHash]
       case MessageBodyType.SUBMIT_TERMS:
-        return [(message.body as ISubmitTerms).content.claim.cTypeHash]
+        // Need to fix with the Message compress and decompress update
```
Comment (tjwelde) [inline src/services/MessageRepository.tsx:250]: What needs to be fixed? the `as IPartialClaim`?

File: src/state/ducks/Claims.ts [hunk 10]
```diff
@@ -416,7 +416,7 @@ const getCTypeHash = (
 
 const getClaimsByCTypeHash = createSelector(
   [getClaims, getCTypeHash],
-  (entries: Entry[], cTypeHash: ICType['cType']['hash']) =>
+  (entries: Entry[], cTypeHash: ICType['ctypeHash']) =>
```
Comment (tjwelde) [inline src/state/ducks/Claims.ts:419]: this `ICType` is not the one coming from the sdk. I can't see `ctypeHash` on it

File: src/state/ducks/Wallet.ts [hunk 11]
```diff
@@ -87,27 +87,26 @@ class Store {
     const identities: { [key: string]: IMyIdentity } = {}
 
     await Promise.all(
```
Comment (tjwelde) [inline src/state/ducks/Wallet.ts:89]: Since the callback function is not async anymore, we would not need `Promise.all` anymore

File: src/services/DelegationsService.ts [hunk 12]
```diff
@@ -172,6 +175,42 @@ class DelegationsService {
 
     return delegation.store(selectedIdentity)
   }
+
+  public static async checkTraversalStepsToParent(
```
Comment (tjwelde) [inline src/services/DelegationsService.ts:179]: Is this the same as in the SDK? If yes, please just use the function from the SDK

**Review by tjwelde — CHANGES_REQUESTED** [review]

File: src/services/AttestationWorkflow.ts (hunk 6 above)
Comment (Dudleyneedham) [reply src/services/AttestationWorkflow.ts:140]: Yes, because we are fetching going deeper till we get to `MessageRepository.send` where we are waiting for a promise to resolve.

File: src/containers/WalletAdd/WalletAdd.tsx (hunk 7 above)
Comment (Dudleyneedham) [reply src/containers/WalletAdd/WalletAdd.tsx:75]: Yes, we are fetching the `DID` and waiting for the promise to resolve.

File: src/components/SelectClaims/SelectClaims.tsx (hunk 8 above)
Comment (Dudleyneedham) [reply src/components/SelectClaims/SelectClaims.tsx:21]: No, I removed it. I guess when I did some changes it auto-suggested this.

File: src/services/MessageRepository.tsx (hunk 9 above)
Comment (Dudleyneedham) [reply src/services/MessageRepository.tsx:249]: Yes, I have done this. I forgot to push.

File: src/state/ducks/Claims.ts (hunk 10 above)
Comment (Dudleyneedham) [reply src/state/ducks/Claims.ts:419]: Yea, I think this is similar to above with the auto-filling when changing something else.

File: src/services/DelegationsService.ts (hunk 12 above)
Comment (Dudleyneedham) [reply src/services/DelegationsService.ts]: Removed this. I had forgotten to push my changes.

File: src/components/DevTools/DevTools.wallet.tsx (hunk 4 above)
Comment (Dudleyneedham) [reply src/components/DevTools/DevTools.wallet.tsx:36]: No, I altered the `BsIdentity.save` so both are no longer async.

File: src/components/MessageDetailView/MessageDetailView.tsx (hunk 5 above)
Comment (Dudleyneedham) [reply src/components/MessageDetailView/MessageDetailView.tsx:131]: I removed the comment.

File: src/components/DevTools/DevTools.wallet.tsx [hunk 13]
```diff
@@ -44,56 +44,52 @@ class BsIdentity {
...
-            },
-            publicIdentity: identity.getPublicIdentity(),
-          }
-          persistentStoreInstance.store.dispatch(
-            Contacts.Store.addContact(newContact)
-          )
+    if (!selectedIdentity) {
+      throw new Error('No Identity selected')
+    }
+    return BalanceUtilities.makeTransfer(
```
Comment (tjwelde) [inline src/components/DevTools/DevTools.wallet.tsx:57]: That function seems wrong. It was resolving the Promise with the `newIdentity` in the callback. Now it is returning whatever `BalanceUtilities.makeTransfer` returns

**Review by tjwelde — CHANGES_REQUESTED** [review]

File: src/services/AttestationWorkflow.ts (hunk 6 above)
Comment (tjwelde) [reply src/services/AttestationWorkflow.ts:140]: But only because you return a function, which itself returns a promise, doesn't mean you have to use `async`.
You have to use async, when you either want to use `await` (which we do not use here anymore), or if you return a promise (which is the case here) and you might throw an error in this function.

Comment (tjwelde) [reply src/services/AttestationWorkflow.ts:140]: The error case is relevant, because someone using this function might use a `.catch()` function on the returned promise, instead of `try...catch`.

File: src/components/SelectClaims/SelectClaims.tsx (hunk 8 above)
Comment (tjwelde) [reply src/components/SelectClaims/SelectClaims.tsx]: You changed `ICType` coming from the demo-client to `ICType` coming from the sdk, which is ok, but unnecessary for this PR.

File: src/state/ducks/Claims.ts (hunk 10 above)
Comment (tjwelde) [reply src/state/ducks/Claims.ts]: I don't know, if we want to change the `ICType` to the one from the SDK. There is no need for this PR to change it

File: src/containers/Tasks/ImportAttestation/ImportAttestation.tsx [hunk 14]
```diff
@@ -49,30 +49,29 @@ const ImportAttestation: React.FC<Props> = ({
...
+            setRequestForAttestation(request)
+            setClaimId(claim.id)
           }
         })
       }
-
-      setRequestForAttestation(newRequestForAttestation)
-      setClaimId(newClaimId)
     })
-  }, [claims, attestation])
```
Comment (tjwelde) [inline src/containers/Tasks/ImportAttestation/ImportAttestation.tsx:75]: Why these changes in this file?

**Review by tjwelde — CHANGES_REQUESTED** [review]

File: src/containers/Tasks/ImportAttestation/ImportAttestation.tsx (hunk 14 above)
Comment (Dudleyneedham) [reply src/containers/Tasks/ImportAttestation/ImportAttestation.tsx:75]: The attestation was being returned as `undefined` when importing. The data was there but not being assigned.

File: src/components/SelectClaims/SelectClaims.tsx (hunk 8 above)
Comment (Dudleyneedham) [reply src/components/SelectClaims/SelectClaims.tsx]: I have reverted it back to its original way.

File: src/state/ducks/Claims.ts (hunk 10 above)
Comment (Dudleyneedham) [reply src/state/ducks/Claims.ts]: I have reverted the changes.

File: src/components/DevTools/DevTools.wallet.tsx (hunk 13 above)
Comment (Dudleyneedham) [reply src/components/DevTools/DevTools.wallet.tsx:57]: I should just revert it, then return a `Promise<void | IMyIdentity>` where it was used.

File: src/containers/Tasks/ImportAttestation/ImportAttestation.tsx (hunk 14 above)
Comment (tjwelde) [reply src/containers/Tasks/ImportAttestation/ImportAttestation.tsx]: we talked about it and reverted it

**Review by tjwelde — APPROVED** [review]
LGTM!
